servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -0,0 +1,136 @@
+                    }
+                }
+                // Either we require a new connection or there wasn't one already established so
+                // try to make a new one if the host is healthy. If it's not healthy, we fail
+                // and let the higher level retries decide what to do.
+                if (!host.isActiveAndHealthy()) {
+                    return noActiveHostsException(hosts);
+                }
+                return host.newConnection(selector, forceNewConnectionAndReserve, context);
+            default:
+                return p2c(size, hosts, getRandom(), selector, forceNewConnectionAndReserve, context);
+        }
+    }
+
+    @Nullable
+    private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random random, Predicate<C> selector,
+                          boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
+        for (int j = maxEffort; j > 0; j--) {
+            // Pick two random indexes that don't collide. Limit the range on the second index to 1 less than
+            // the max value so that if there is a collision we can safety increment. We also increment if
+            // i2 > i1 to avoid biased toward lower numbers since we limited the range by 1.
+            final int i1 = random.nextInt(size);
+            int i2 = random.nextInt(size - 1);
+            if (i2 >= i1) {
+                ++i2;
+            }
+            Host<ResolvedAddress, C> t1 = hosts.get(i1);
+            Host<ResolvedAddress, C> t2 = hosts.get(i2);
+            // Make t1 the preferred host by score to make the logic below a bit cleaner.
+            if (t1.score() < t2.score()) {
+                Host<ResolvedAddress, C> tmp = t1;
+                t1 = t2;
+                t2 = tmp;
+            }
+
+            if (!forceNewConnectionAndReserve) {
+                // First we're going to see if we can get an existing connection regardless of health status. Since t1
+                // is 'better' we'll try it first. If it doesn't have any existing connections we don't fall back to t2
+                // or else we would cause a bias toward hosts with existing connections which could ultimately drive all
+                // traffic to the first host to make a connection in the case of a multiplexed session.
+                C c = t1.pickConnection(selector, contextMap);
+                if (c != null) {
+                    return succeeded(c);
+                }
+                // We now need to consider the health status and make a new connection if either
+                // host is considered healthy.
+            }
+
+            // We either couldn't find a live connection or are being forced to make a new one. Either way we're
+            // going to make a new connection which means we need to consider health.
+            final boolean t1Healthy = t1.isActiveAndHealthy();
+            final boolean t2Healthy = t2.isActiveAndHealthy();
```

Comment:
Right now this follows the pattern set by the `RoundRobinSelector` which is to consider health only for making new connections. I think that is born of the only current circuit breaker being the L4 consecutive failures and the current hesitancy to mark a connection as inactive.
I think we should change that behavior in that we should consider health for any connections and also allow an unhealthy connection to be selected as a last resort (likely configurable). However, that is a reasonably big reinterpretation of what health means. I'd like to change it for RR and P2C at the same time, or maybe switch RR first and then merge this. Feedback appreciated on that.

Reply:
I agree. I think we can potentially go further than RR and offer a panic mode, maybe in two flavors: existing connections only and even for new connections, that at least lets the LB try. At least for the latter it's probably important to add some restrictions at the connection factory level to avoid connection storms so we should proceed with caution there.

Reply:
+1 for improving and interating on this area and shaping definition of "unhealthy" state.

> if we can't connect to a host there is a good chance existing connections are either not responsive or draining

In order to keep track of use-cases and make sure we don't miss anything, it's worth mentioning that `ServerListenContext` API allows users to pause acceptance of new connections without affecting pre-existing connections. We should account for that as well.